sqlmesh/core/macros.py

```diff
@@ -648,6 +648,87 @@ def eval_(evaluator: MacroEvaluator, condition: exp.Condition) -> t.Any:
     return evaluator.eval_expression(condition)
 
 
+@macro()
+def star(
+    evaluator: MacroEvaluator,
+    relation: exp.Table,
+    alias: t.Optional[exp.Identifier | exp.Column] = None,
+    except_: t.Optional[exp.Array | exp.Tuple] = None,
+    prefix: t.Optional[exp.Literal] = None,
+    suffix: t.Optional[exp.Literal] = None,
+    quote_identifiers: exp.Boolean = exp.true(),
+) -> t.List[exp.Expression]:
+    """Returns a list of projections for the given relation.
+
+    Example:
+        >>> from sqlglot import parse_one
+        >>> from sqlmesh.core.macros import MacroEvaluator
+        >>> sql = "SELECT @STAR(foo, bar, [c], 'baz_') FROM foo AS bar"
+        >>> MacroEvaluator(schema={"foo": {"a": "string", "b": "string", "c": "string", "d": "int"}}).transform(parse_one(sql)).sql()
+        'SELECT CAST("bar"."a" AS TEXT) AS "baz_a", CAST("bar"."b" AS TEXT) AS "baz_b", CAST("bar"."d" AS INT) AS "baz_d" FROM foo AS bar'
+    """
+    if alias and not isinstance(alias, (exp.Identifier, exp.Column)):
+        raise SQLMeshError(f"Invalid alias '{alias}'. Expected an identifier.")
+    if except_ and not isinstance(except_, (exp.Array, exp.Tuple)):
+        raise SQLMeshError(f"Invalid except '{except_}'. Expected an array.")
+    if prefix and not isinstance(prefix, exp.Literal):
+        raise SQLMeshError(f"Invalid prefix '{prefix}'. Expected a literal.")
+    if suffix and not isinstance(suffix, exp.Literal):
+        raise SQLMeshError(f"Invalid suffix '{suffix}'. Expected a literal.")
+    if not isinstance(quote_identifiers, exp.Boolean):
+        raise SQLMeshError(
+            f"Invalid quote_identifiers '{quote_identifiers}'. Expected a boolean."
+        )
+    projections = []
+    exclude = []
+    kwargs = {"quoted": quote_identifiers.this}
+    if alias:
+        kwargs["table"] = alias.name
+    if except_ and isinstance(except_.expressions, list):
+        exclude.extend(
+            e.name
+            for e in except_.expressions
+            if isinstance(e, (exp.Identifier, exp.Column))
+        )
+    for column, type_ in evaluator.columns_to_types(relation.sql()).items():
+        if column in exclude:
+            continue
+        p = prefix.this if prefix else ""
+        s = suffix.this if suffix else ""
+        projections.append(
+            exp.cast(exp.column(column, **kwargs), type_).as_(
+                f"{p}{column}{s}", quoted=kwargs["quoted"]
+            )
+        )
+    return projections
+
+
+@macro()
+def generate_surrogate_key(
+    _: MacroEvaluator, *fields: exp.Column | exp.Identifier
+) -> exp.Func:
```

Comment:
@eakmanrq do you have opinions on this

Reply:
Seems fine. It is an opinionated way to generate surrogate keys but if we add arguments like the hash function to use then it somewhat defeats the purpose since now it may not be implemented consistently across an org. Also this has an opinionated constant for null but yet again it needs to be consist so it makes sense to now allow overriding. I think for larger orgs they will implement their own with their own opinions but this can be used as a reference. 

If you are asking because of past conversations about creating new columns by concating that was specific to doing that automatically within joins and therefore forcing shuffles. Doesn't apply here.